Remora.Rest/Json/Internal/BoundDataObjectConverter.cs

```diff
@@ -0,0 +1,272 @@
+
+    /// <inheritdoc />
+    public override bool CanConvert(Type typeToConvert)
+    {
+        return typeToConvert == typeof(TImplementation) || base.CanConvert(typeToConvert);
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="BoundDataObjectConverter{TInterface, TImplementation}"/> class.
+    /// </summary>
+    /// <param name="source">The source converter factory that created this instance.</param>
+    /// <param name="options">The serializer options this instance is bound to.</param>
+    public BoundDataObjectConverter(DataObjectConverter<TInterface, TImplementation> source, JsonSerializerOptions options)
+    {
+        _dtoFactory = source.DTOFactory;
+        _allowExtraProperties = source.DoesAllowExtraProperties;
+
+        var dtoProperties = source.DTOProperties;
+
+        var writeProperties = new List<DTOPropertyInfo>();
+        var readProperties = new List<DTOPropertyInfo>();
+
+        var interfaceMap = typeof(TImplementation).GetInterfaceMap(typeof(TInterface));
+
+        foreach (var property in dtoProperties)
+        {
+            var converter = source.GetConverter(property, options);
+            var propertyOptions = GetPropertyConverterOptions(options, converter);
+            var defaultValue = source.GetDefaultValueForType(property.PropertyType);
+            var readNames = source.GetReadJsonPropertyName(property, options);
+            var writeNames = source.GetWriteJsonPropertyName(property, options);
+            var write = source.GetPropertyWriter(property);
+
+            // We cache this as well since the check is somewhat complex
+            bool allowsNull = property.AllowsNull();
+
+            var data = new DTOPropertyInfo
+            (
+                property,
+                readNames,
+                writeNames,
+                write,
+                allowsNull,
+                defaultValue,
+                propertyOptions,
+                readProperties.Count
+            );
+
+            if (property.CanWrite)
+            {
+                // If a property is writable, it can be *read* from JSON.
+                readProperties.Add(data);
+            }
+
+            if (property.CanWrite || source.IncludesReadOnlyProperty(property))
+            {
+                // Any property that is writable and not excluded due to being read-only,
+                // can be *written* to JSON.
+                writeProperties.Add(data);
+            }
+        }
+
+        _writeProperties = writeProperties.ToArray();
+        _readProperties = readProperties.ToArray();
+
+        _readPropertiesByName = readProperties
+            .SelectMany(p => p.ReadNames.Select((n, i) => (IsPrimary: i == 0, n, DTOProperty: p)))
+            .ToDictionary(x => x.n, x => (x.IsPrimary, x.DTOProperty));
+    }
+
+    private static JsonSerializerOptions GetPropertyConverterOptions(JsonSerializerOptions options, JsonConverter? converter)
```

Comment:
`CreatePropertyConverterOptions`  might be more appropriate, since the method allocates a new instance and doesn't merely access data.